xlsynth-g8r/src/equiv/solver_interface.rs

```diff
@@ -0,0 +1,851 @@
+        let a = solver.declare("a", 8).unwrap();
+        let b = solver.numerical(16, 0x89ab);
+        let c = solver.extract(&b, 12, 5);
+        let eq_constraint = solver.eq(&a, &c);
+        solver.assert(&eq_constraint).unwrap();
+        assert_eq!(solver.check().unwrap(), Response::Sat);
+        assert_eq!(
+            solver.get_value(&a, &ir::Type::Bits(8)).unwrap(),
+            IrValue::make_ubits(8, 0x4d).unwrap()
+        );
+    }
+
+    pub fn test_slice<S: Solver>(solver: &mut S) {
+        let a = solver.declare("a", 8).unwrap();
+        let b = solver.numerical(16, 0x89ab);
+        let c = solver.slice(&b, 5, 8);
+        let eq_constraint = solver.eq(&a, &c);
+        solver.assert(&eq_constraint).unwrap();
+        assert_eq!(solver.check().unwrap(), Response::Sat);
+        assert_eq!(
+            solver.get_value(&a, &ir::Type::Bits(8)).unwrap(),
+            IrValue::make_ubits(8, 0x4d).unwrap()
+        );
+    }
+
+    pub fn assert_solver_eq<S: Solver>(solver: &mut S, lhs: &BitVec<S::Rep>, rhs: &BitVec<S::Rep>) {
+        solver.push().unwrap();
+        let eq_constraint = solver.eq(&lhs, &rhs);
+        solver.assert(&eq_constraint).unwrap();
+        assert_eq!(solver.check().unwrap(), Response::Sat);
+        solver.pop().unwrap();
+        solver.push().unwrap();
+        let ne_constraint = solver.ne(&lhs, &rhs);
+        solver.assert(&ne_constraint).unwrap();
+        assert_eq!(solver.check().unwrap(), Response::Unsat);
+        solver.pop().unwrap();
+    }
+
+    #[macro_export]
+    macro_rules! test_get_value {
+        ($fn_name:ident,
+         $solver:expr,
+         $in_w:expr,  $in_val:expr,
+         $out_ty:expr, $out_val:expr $(,)?
+        ) => {
+            #[test]
+            fn $fn_name() {
+                let mut solver = $solver;
+                let operand = solver.from_raw_str($in_w, $in_val);
+                solver.check().unwrap();
+                let actual = solver.get_value(&operand, &$out_ty).unwrap();
+                let expected = $out_val;
+                assert_eq!(actual, expected);
+            }
+        };
+    }
+
+    // -----------------------------------------------------------------------
+    // Generic "one-operand → one result" test helper for Solver.
+    //
+    //  • `$fn_name`     – the Rust test-function identifier that will be
+    //                     generated by the macro.
+    //  • `$op`          – the `Solver` method to call (e.g. `not`, `neg`,
+    //                     `reverse`, `or_reduce`, …).
+    //  • `$in_w`        – bit-width of the operand fed to the solver.
+    //  • `$in_val`      – literal value of that operand.
+    //  • `$out_w`       – bit-width expected from the operation
+    //                     (use 1 for reductions).
+    //  • `$out_val`     – literal value expected from the operation.
+    // -----------------------------------------------------------------------
+    #[macro_export]
```

Comment:
We may want to make a dedicated test file under tests/ to avoid the macro overhead affecting the compile times for the crate, or at least marking the test_utils module with #[cfg(test)] to have it only build when we're in testing mode.